zebrad/src/components/mempool/crawler.rs

```diff
@@ -0,0 +1,88 @@
+//! Zebra Mempool crawler.
+//!
+//! The crawler periodically requests transactions from peers in order to populate the mempool.
+
+use std::time::Duration;
+
+use futures::{stream, StreamExt, TryStreamExt};
+use tokio::{sync::Mutex, time::sleep};
+use tower::{BoxError, Service, ServiceExt};
+
+use zebra_network::{Request, Response};
+
+/// The number of peers to request transactions from per crawl event.
+const FANOUT: usize = 4;
+
+/// The delay between crawl events.
+const RATE_LIMIT_DELAY: Duration = Duration::from_secs(75);
+
+/// The mempool transaction crawler.
+pub struct Crawler<S> {
+    peer_set: Mutex<S>,
+}
+
+impl<S> Crawler<S>
+where
+    S: Service<Request, Response = Response, Error = BoxError> + Clone + Send + 'static,
+    S::Future: Send,
+{
+    /// Spawn an asynchronous task to run the mempool crawler.
+    pub fn spawn(peer_set: S) {
+        let crawler = Crawler {
+            peer_set: Mutex::new(peer_set),
+        };
+
+        tokio::spawn(crawler.run());
+    }
+
+    /// Periodically crawl peers for transactions to include in the mempool.
+    pub async fn run(self) {
+        loop {
+            self.wait_until_enabled().await;
+            self.crawl_transactions().await;
+            sleep(RATE_LIMIT_DELAY).await;
+        }
+    }
+
+    /// Wait until the mempool is enabled.
+    async fn wait_until_enabled(&self) {
+        // TODO: Check if synchronizing up to chain tip has finished (#).
+    }
+
+    /// Crawl peers for transactions.
+    ///
+    /// Concurrently request [`FANOUT`] peers for transactions to include in the mempool.
+    async fn crawl_transactions(&self) {
+        let requests = stream::repeat(Request::MempoolTransactionIds).take(FANOUT);
+        let peer_set = self.peer_set.lock().await.clone();
+
+        trace!("Crawling for mempool transactions");
+
+        let result = peer_set
+            .call_all(requests)
+            .unordered()
+            .try_for_each(|response| self.handle_response(response))
+            .await;
+
+        if let Err(error) = result {
+            error!("Failed to crawl peers for mempool transactions: {}", error);
```

Comment:
Nit: `error!` is a very loud log message.

We have redundant requests via the `FANOUT`, so we could just ignore the error.
But we probably want `info!` while we're developing the mempool, so repeated failures are obvious.
```suggestion
            info!("Failed to crawl peers for mempool transactions: {}", error);
```

Reply:
Done ([6a2f952](https://github.com/jvff/zebra/commit/6a2f95233b96405a77fc81a44643035c7abc81ae)), and I opened #2655 to reduce it more in the future.